# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/indexedDB/IndexedDatabaseManager.cpp
# Commit: bb094a6084b3
# Full Hash: bb094a6084b3b127c090827ee8ede4b06cdbdea6
# Author: Simon Giesecke <sgiesecke@mozilla.com>
# Date: 2020-03-19 17:04:52
# Regressor Bug: 1623278
# File Overlap Count: 1
# Description:
#   Bug 1623278 - Use SafeRefPtr for FileManager. r=dom-workers-and-storage-reviewers,asuth
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D65549
# ==============================================================================

diff -r 7fdf38a403bb -r bb094a6084b3 dom/indexedDB/IndexedDatabaseManager.cpp
--- a/dom/indexedDB/IndexedDatabaseManager.cpp	Thu Mar 19 09:32:02 2020 +0000
+++ b/dom/indexedDB/IndexedDatabaseManager.cpp	Thu Mar 19 09:32:53 2020 +0000
@@ -60,10 +60,10 @@
 
 class FileManagerInfo {
  public:
-  MOZ_MUST_USE RefPtr<FileManager> GetFileManager(
+  MOZ_MUST_USE SafeRefPtr<FileManager> GetFileManager(
       PersistenceType aPersistenceType, const nsAString& aName) const;
 
-  void AddFileManager(FileManager* aFileManager);
+  void AddFileManager(SafeRefPtr<FileManager> aFileManager);
 
   bool HasFileManagers() const {
     AssertIsOnIOThread();
@@ -81,16 +81,17 @@
                                       const nsAString& aName);
 
  private:
-  nsTArray<RefPtr<FileManager> >& GetArray(PersistenceType aPersistenceType);
+  nsTArray<SafeRefPtr<FileManager> >& GetArray(
+      PersistenceType aPersistenceType);
 
-  const nsTArray<RefPtr<FileManager> >& GetImmutableArray(
+  const nsTArray<SafeRefPtr<FileManager> >& GetImmutableArray(
       PersistenceType aPersistenceType) const {
     return const_cast<FileManagerInfo*>(this)->GetArray(aPersistenceType);
   }
 
-  nsTArray<RefPtr<FileManager> > mPersistentStorageFileManagers;
-  nsTArray<RefPtr<FileManager> > mTemporaryStorageFileManagers;
-  nsTArray<RefPtr<FileManager> > mDefaultStorageFileManagers;
+  nsTArray<SafeRefPtr<FileManager> > mPersistentStorageFileManagers;
+  nsTArray<SafeRefPtr<FileManager> > mTemporaryStorageFileManagers;
+  nsTArray<SafeRefPtr<FileManager> > mDefaultStorageFileManagers;
 };
 
 }  // namespace indexedDB
@@ -679,7 +680,7 @@
   mBackgroundActor = nullptr;
 }
 
-RefPtr<FileManager> IndexedDatabaseManager::GetFileManager(
+SafeRefPtr<FileManager> IndexedDatabaseManager::GetFileManager(
     PersistenceType aPersistenceType, const nsACString& aOrigin,
     const nsAString& aDatabaseName) {
   AssertIsOnIOThread();
@@ -692,7 +693,8 @@
   return info->GetFileManager(aPersistenceType, aDatabaseName);
 }
 
-void IndexedDatabaseManager::AddFileManager(FileManager* aFileManager) {
+void IndexedDatabaseManager::AddFileManager(
+    SafeRefPtr<FileManager> aFileManager) {
   AssertIsOnIOThread();
   NS_ASSERTION(aFileManager, "Null file manager!");
 
@@ -702,7 +704,7 @@
     mFileManagerInfos.Put(aFileManager->Origin(), info);
   }
 
-  info->AddFileManager(aFileManager);
+  info->AddFileManager(std::move(aFileManager));
 }
 
 void IndexedDatabaseManager::InvalidateAllFileManagers() {
@@ -855,7 +857,7 @@
   return idbManager->mLocale;
 }
 
-RefPtr<FileManager> FileManagerInfo::GetFileManager(
+SafeRefPtr<FileManager> FileManagerInfo::GetFileManager(
     PersistenceType aPersistenceType, const nsAString& aName) const {
   AssertIsOnIOThread();
 
@@ -865,17 +867,17 @@
   const auto foundIt =
       std::find_if(managers.cbegin(), end, DatabaseNameMatchPredicate(&aName));
 
-  return foundIt != end ? *foundIt : nullptr;
+  return foundIt != end ? foundIt->clonePtr() : nullptr;
 }
 
-void FileManagerInfo::AddFileManager(FileManager* aFileManager) {
+void FileManagerInfo::AddFileManager(SafeRefPtr<FileManager> aFileManager) {
   AssertIsOnIOThread();
 
-  nsTArray<RefPtr<FileManager> >& managers = GetArray(aFileManager->Type());
+  nsTArray<SafeRefPtr<FileManager> >& managers = GetArray(aFileManager->Type());
 
   NS_ASSERTION(!managers.Contains(aFileManager), "Adding more than once?!");
 
-  managers.AppendElement(aFileManager);
+  managers.AppendElement(std::move(aFileManager));
 }
 
 void FileManagerInfo::InvalidateAllFileManagers() const {
@@ -900,7 +902,7 @@
     PersistenceType aPersistenceType) {
   AssertIsOnIOThread();
 
-  nsTArray<RefPtr<FileManager> >& managers = GetArray(aPersistenceType);
+  nsTArray<SafeRefPtr<FileManager> >& managers = GetArray(aPersistenceType);
 
   for (uint32_t i = 0; i < managers.Length(); i++) {
     managers[i]->Invalidate();
@@ -924,7 +926,7 @@
   }
 }
 
-nsTArray<RefPtr<FileManager> >& FileManagerInfo::GetArray(
+nsTArray<SafeRefPtr<FileManager> >& FileManagerInfo::GetArray(
     PersistenceType aPersistenceType) {
   switch (aPersistenceType) {
     case PERSISTENCE_TYPE_PERSISTENT: